### ISSUE TITLE
Let's talk about fragments

### ISSUE THREAD

**Issue by rianby64** [m3co/pseudoimport-html#64]
@TakayoshiKochi hello. I'm very happy to see you here!

Here I'd like to answer to [your initial questions](https://github.com/TakayoshiKochi/script-type-module/issues/1#issuecomment-285586544)

> I didn't have clarified what this repository (branch from original ES modules polyfill) is for...

This repo is an attempt to write fragments that can work from IE9+. (It is not a polyfill for ES modules) It's an attempt to organize the code to avoid difficult things related to initialization, bundling and other stages related to the development of web-apps. I.e., Angular 1 forces you to package "directives" into difficult function-wrappers. Angular 2 forces you to use something called ```@Component```, React has its own JSX that mixes HTML with JS... and so on. It looks like all frameworks de-facto restrict your code to be into a frame. And I'm convinced that the browser is the framework itself.

> So it sounds like there are some commonality with the fragments

I often ask if somebody has already developed this idea... but I couldn't find anything similar :( I will be very grateful if you could give me a hint.

Now, the questions :smile: 
> 1. Once the content of HTML is loaded, will it be attached under the element?

Yes, it will. By default, it's done via [appendChild(html)](https://github.com/m3co/pseudoimport-html/blob/4511443b02f3d49b16dd99838290bad165d94933/src/fragment-ce.js#L149). However, I'd like [to retake insertAdjacentHTML](https://github.com/m3co/pseudoimport-html/issues/30) in the future...

> 2. If the content is attached then <script> is executed, <style> is applied, or to avoid the side-effect, does it keep the HTML snippet within some DocumentFragment (as the name suggests)?

Good point. I solved that using [DocumentFragment](https://github.com/m3co/pseudoimport-html/blob/4511443b02f3d49b16dd99838290bad165d94933/includes/utils.js#L66) but I ran into some problems. First of all, the ```createContextualFragment``` function doesn't work in old browsers, so I was forced to write [my version](https://github.com/m3co/pseudoimport-html/blob/4511443b02f3d49b16dd99838290bad165d94933/includes/utils.js#L63) called ```craftedCreateContextualFragment```
The second problem is that because of ```document.currentFragment``` I was forced to [turn all the scripts into inline scripts](https://github.com/m3co/pseudoimport-html/blob/4511443b02f3d49b16dd99838290bad165d94933/includes/utils.js#L98) :( and this looks weird. If all browsers would have implemented ```beforescriptexecute``` event... then I could perform the fragments in a more elegant way.

> 3. Can a fragment have dependency? I.e. if a loaded fragment has <x-fragment> inside, this sub fragment will be processed accordingly?

Sub-fragments are processed according to their appearance. Each fragment, once loaded, it fires an event ```load```, but in reversed order. I mean, a parent fragment will fetch all its children-fragments, each of them will fire its ```load``` event and once all children have been loaded then the parent will fire its ```load``` event. Another point that I handle is the presence of circular dependencies. When a circular dependency is found then I throw an error an stop the processing of that particular circular fragment.

All ```src``` attributes are pre-processed in order to handle _relative paths_. This concerns ```<x-fragment src="...">```[(1)](https://github.com/m3co/pseudoimport-html/blob/master/src/fragment-ce.js#L49) and ```<script src="..."></script>```[(2)](https://github.com/m3co/pseudoimport-html/blob/master/includes/utils.js#L89).

While I was writing this letter to you, I found that would be great if I could manage the value of ```<base>```. As far as I know, it's a read-only value. More than one ```<base>``` is not allowed in the document. Because of that, I see that all the ```fetch(url)``` must be absolute paths :(
Maybe will be great to offer something like ```Fetch.baseURL = "base-url"```?

Here's a demo: http://container3.m3c.space/demo/

@TakayoshiKochi, thanks a lot!

**Comment by rianby64** [comment]
I'm also convinced that ```<link rel="import" href="url" />``` offers a nice solution. But it's a shame that [Firefox will not ship it](https://developer.mozilla.org/en-US/docs/Web/Web_Components/HTML_Imports).

**Comment by TakayoshiKochi** [comment]
Hi, thanks for your answers!

Yeah, the bigger problem we have is to provide an alternative to HTML imports without losing the benefits of it as much as possible.

For base URLs, for fetching you can use `new URL('relative_path', document.baseURI).toString()`?

**Comment by rianby64** [comment]
@TakayoshiKochi I'm happy to see you here ))

Ok, here we go: https://jsfiddle.net/rianby64/zjuo6qox/8/
1. Let's imagine that the ```<my-custom-element></my-custom-element>``` will not do modifications to the fetched HTML code. It will behave like a _proxy_ as fragments do.
2. Then, the imported code may content some _fetchs_ that have relative paths. In my case, they're absolute :(
3. How to implicit set the ```new URL('relative_path', document.baseURI).toString()``` without modifying the fetched-imported HTML code? I mean, how to setup a baseURI for all fetch calls?

What I would like to solve here, is to have a mechanism that before doing ```appendChild``` it would setup "a global baseURI". In that way, [a fetch like this one](https://gist.github.com/rianby64/fb7f9d147ade5436402a2ebf232856be#file-mycetmpl-html-L6) may stay relative.

**Comment by rianby64** [comment]
I've another idea that may be useful. It's related to ```document.currentScript```.
I know that there's a discussion around that variable. And there's an intention not to implement ```document.currentScript```. But, in the other hand, I'd like to be able to understand the current place of my script. And usually it's done by writing something like
```javascript
'use strict';
var ownerDocument = document.currentScript.ownerDocument;
```

But what about if use something like
```javascript
var ownerDocument = HTMLImport.currentDocument;
```
?

For fragments I use ```document.currentFragment```.

**Comment by rianby64** [comment]
@TakayoshiKochi I'd be happy to keep this conversation at IRC or here. Please, let's talk about this topic ```https : // github.com/ w3c/webcomponents/issues/645```

**Comment by TakayoshiKochi** [comment]
Hi, sorry for not being here for long time.  What do you want me to discuss here (or if IRC, what channel are you in)?
If you are interested in HTML Modules design itself, please feel free to join the discussion there.

**Comment by rianby64** [comment]
I'm at ``IRC freenode #whatwg``` and my nickname is ```rianby64```

**Comment by rianby64** [comment]
Basically, I'd like to explain to you some things that I found while developing an instrument for my own projects. It looks pretty similar to your tag ```<include src="...">```. Why not there in the public IRC? Because I feel not confident enough...

**Comment by TakayoshiKochi** [comment]
Sorry, I don't have infinite amount of time for projects that I'm working on, I don't have time to advice specifically for your project myself.

The `<include>` tag, as I put as an example, is probably a separate project from what HTML Modules will be, so feel free to put your proposal in https://github.com/whatwg/html/ .  Don't be shy.

To add some more context, we (as a broader meaning, that people who develop web standards and implement them in browsers) are interested in what are missing in the platform for web developers or users. If something is very commonly used as a polyfill (e.g. imagine jQuery of old days, now we have `querySelector()` et al.), we might prioritize to standardize to have one in common in all browsers, but we may not, if that could be polyfilled with existing primitives well enough, that borderline is not so clear.

So a feature such as `<include>`, which could be polyfilled via `fetch()` and [DOM Parser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) to some extent, there should be a strong reason to exist in HTML standards, which is a hard part to argue with every implementer.

Good Luck!

**Comment by rianby64** [comment]
@TakayoshiKochi Thanks a lot for this advice. I'll try to use more cases and spread the word. Maybe the idea around ```<include src="...">``` worths more brainstorming. But for me, I've some examples that show the advantages if use this tag. For instance, I can split my app in small pieces. Thanks a lot again.

**Comment by warpech** [comment]
@tomalec and I (mostly @tomalec) have implemented a similar solution using HTML Imports spec: https://github.com/Juicy/imported-template

The method that we use is explained in detail here: https://starcounter.io/html-partialsincludes-webcomponents-way/

TL;DR: Our method relies on stamping the first `<template>` found in the imported HTML document. Everything outside of that template is resolved only once, including:

- `<link rel="stylesheet">`
- `<script>`

This is a huge win because you have the document fragment that can be stamped multiple times as well as setup fragments that are executed only once.

**Comment by tomalec** [comment]
TL;DR: Our method relies on stamping the **root** `<template>` element**s**  found in the imported HTML document. Everything outside of that template is resolved **as usual in document imported via HTML Import** - only once, including:

- `<link>`, ` rel="import"`, ` rel="stylesheet"`
- `<script>`
- `<a-custom-or-native-element>`
- ...

Also the root `template` could contain any valid HTML Element that will be stamped many times.
This is a huge win because you have the document fragment that can be stamped multiple times as well as setup fragments that are executed only once.

**Comment by rianby64** [comment]
Nice approach :+1:  I see our solutions differ on the way to handle the imported. Resolve only once and reuse the imported many times, or, Resolve every time after importing. What are the implications from both approaches? For me is a good question. Also, looks like your approach, as HTML import does, is the correct broad-adopted way.

In the other hand,
```
<html>
  ...
  <x-fragment src="an-URL.html"></x-fragment>
</html>
```
, express the idea that the ```x-fragment``` is importing some HTML stuff... That looks like: fire-and-forget.

@tomalec , @warpech I'll happy to chat with you about these ideas. Thanks a lot!

**Comment by rianby64** [comment]
IRC: ```freenode #htmlimport```
